apps/aecore/src/aec_pow_cuckoo.erl

```diff
@@ -141,49 +142,105 @@ generate_single(Hash, Nonce, Target) ->
+-define(POW_TOO_BIG, {error, nonce_too_big}).
+-define(POW_TOO_SMALL, {error, nonces_not_ascending}).
+-define(POW_NON_MATCHING, {error, endpoints_do_not_match_up}).
+-define(POW_BRANCH, {error, branch_in_cycle}).
+-define(POW_DEAD_END, {error, cycle_dead_ends}).
+-define(POW_SHORT_CYCLE, {error, cycle_too_short}).
+-define(PROOFSIZE, 42).
 %%------------------------------------------------------------------------------
 %% @doc
-%%   Proof of Work verification (without difficulty check)
-%%
-%%   This function returns {error, term()} in the unlikely case that
-%%   it does not manage to determine an outcome for the verification
-%%   of the PoW (e.g. verifier terminates abnormally).
+%%   Proof of Work verification (difficulty check should be done before calling
+%%   this function)
 %% @end
 %%------------------------------------------------------------------------------
--spec verify(Hash :: string(), Nonce :: integer(),
-             Soln :: aec_pow:pow_evidence()) -> boolean().
-verify(Hash, Nonce, Soln) ->
-    BinDir = aecuckoo:bin_dir(),
-    {_, _, Size} = application:get_env(aecore, aec_pow_cuckoo, ?DEFAULT_CUCKOO_ENV),
-    SolnStr = solution_to_hex(Soln),
-    ?info("Executing: ~p~n", [lists:concat(export_ld_lib_path() ++ ["./verify", Size, " -h ", Hash, " -n ", Nonce])]),
-    try exec:run(
-              lists:concat(export_ld_lib_path() ++
-                               ["./verify", Size, " -h ", Hash, " -n ", Nonce]),
-              [{stdout, self()},
-               {stderr, self()},
-               stdin,
-               {kill_timeout, 1}, %% Kills exe with SIGTERM, then with SIGKILL if needed
-               {cd, BinDir},
-               {env, [{"SHELL", "/bin/sh"}]},
-               monitor]) of
-        {ok, _ErlPid, OsPid} ->
-            exec:send(OsPid, list_to_binary(SolnStr)),
-            exec:send(OsPid, eof),
-            ParserState = #state{os_pid = OsPid,
-                                        buffer = [],
-                                        parser = fun parse_verification_result/2},
-            case wait_for_result(ParserState) of
-                {ok, Value} when is_boolean(Value) ->
-                    Value;
-                {error, _Reason} = Err ->
-                    Err
-            end
+-spec verify_proof(Hash :: binary(), Nonce :: integer(),
+                   Solution :: aec_pow:pow_evidence()) -> boolean().
+verify_proof(Hash, Nonce, Solution) ->
+    {_Exe, _Extra, Size} = application:get_env(aecore, aec_pow_cuckoo, ?DEFAULT_CUCKOO_ENV),
+    {K0, K1} = aec_siphash:create_keypair(Hash, Nonce),
+    EdgeMask = (1 bsl (Size - 1)) - 1,
+    try
+        %% Generate Uv pairs representing endpoints by hashing the proof
+        %% XOR points together: for a closed cycle they must match somewhere
+        %% making one of the XORs zero.
+        {Xor0, Xor1, _, Uvs} =
+            lists:foldl(
+              fun(N, _) when N > EdgeMask ->
+                      throw(?POW_TOO_BIG);
```

Comment:
Is throw/catch used only for quitting early this foldl? Removing the catch would ease reasoning about the code for developers (and for tools like dialyzer).

Reply:
I inherited this part from Akos, didn't really think of the style but I think you have a point that the code could be easier to follow without the throw/catch.